### ISSUE TITLE
Make radar position error configurable per emitter

### ISSUE THREAD

**Issue by WatchTheFort** [beyond-all-reason/spring#1422]
Currently, radar wobble is by default a global setting, controlled by the `isTargetingUpgrade` unit def tag, and the `SetRadarErrorParams` callout. It can also be configured per-unit with the `SetUnitPosErrorParams` callout.

The desire is to base radar wobble on the radar coverage of the specific unit, where position error parameters can be set per emitter. When multiple emitters cover the same unit, a method would be required to specify whether the wobble should be based on the maximum individual emitter strength, or the combined individual emitter strengths.

**Proposal**

- Add new unit def tags, `radarError` and `sonarError`, integer of how much radar wobble to apply
- If unspecified, use the default value as exists in current functionality
- Add new modrule, `BaseRadarError`, integer for default amount of radar wobble
- Add new modrule, `CombinedRadarErrorReduction`, string with values `maximum`, `additive`
- Calculate radar wobble for each radar blip based on the strength of the emitters detecting that blip, before applying any effects from `isTargetingUpgrade`

**Comment by sprunk** [comment]
> Add new modrule, BaseRadarError, integer for default amount of radar wobble

Make a 1-liner gadget.

> Add new modrule, CombinedRadarErrorReduction, string with values maximum, additive

bool?

> Add new unit def tags, `radarError` and `sonarError`, integer of how much radar wobble to apply

Names would ideally be something like `radarErrorReductionLevel`.

**Comment by marcushutchings** [comment]
I definately vouch for the combined strength approach over using a max function. It avoids the need to walk lists when a RADAR zone is abandoned by the current strongest unit. The difference in performance would be significant - especially we are targtting larger scale games than most RTSes.

**Comment by WatchTheFort** [comment]
Sounds good, for our desire of making T1 radars have wobble and T2 radars not, we would just make the the T2 value >> T1 value

**Comment by ChrisFloofyKitsune** [comment]
To combat spamming a bunch of T2 radars in a box of walls, we could have a gadget such that wobble reduction is adjusted to be weaker for a T2 radar based on how many other T2 radars are nearby

in other words, make T2 radars interfere with each other's ability to reduce radar wobble

**Comment by WatchTheFort** [comment]
@ChrisFloofyKitsune That's not the gameplay design we're aiming for, so we'll just make T2 radars strong enough that a single one is sufficient, and adding more will have no discernable effect.

**Comment by marcushutchings** [comment]
While a nice idea. I think we need to try and keep RADARs as simple a reasonably possible. RADAR is a minor feature of these games. I feel the only reason we're have this discussion is because the RADAR wobble feature was added a long time ago as a random idea and we have got used to it being there rather than wanting it specifically. If it wasn't there in the first place, I don't believe any games currently on the engine would be requesting the feature.

**Comment by GoogleFrog** [comment]
> If it wasn't there in the first place, I don't believe any games currently on the engine would be requesting the feature.

I disagree. There would be some design problems to solve if radar wobble didn't exist. Radar wobble isn't the only solution so we wouldn't necessarily settle on it, but I think it is a particularly good solution.

**Comment by KyleAnthonyShepherd** [comment]
Driving by with my 2 cents. Would it be possible to add a callin in `CUnit::UpdatePosErrorParams` so the engine can ask the unit "hey, I'm about to do the radar wobble, do you have any code to run? Return true if I should do default stuff, return 0 if you don't want me to do default wobble"

And let gameside Lua figure out the nearby emitters and how to adjust the error parameters.

**Comment by sprunk** [comment]
IIRC you can set the next wobble to happen after INT_MAX frames and then replicate engine wobble by just polling each unit every N frames.

**Comment by KyleAnthonyShepherd** [comment]
With some clever management of the units to poll using `UnitEnteredRadar`, that seems doable. 

Hard part is I'm not sure if lua got the ability to call into the engine to run raytraces yet, so quickly determining line of sight to T2 radar could still be iffy.

**Comment by KyleAnthonyShepherd** [comment]
`GetUnitWeaponHaveFreeLineOfFire` should work, if we give T2 radars a weapon with the correct collision flags

**Comment by sprunk** [comment]
No raytraces yet but there's a WIP at https://github.com/beyond-all-reason/RecoilEngine/pull/1624, feel free to take it over.
